### PR TITLE
WIP Phrase Entity

### PR DESCRIPTION
**Pull request by jackmakiyama**

### CODE REVIEW

File: src/Phrases/Persistance/Relational.php [hunk 1]
```diff
@@ -42,7 +43,8 @@ protected function getPdo()
     {
         $findOneStatement = $this->createFindOneRandomStatement();
         $findOneStatement->execute();
-        $onePhrase = $findOneStatement->fetch(PDO::FETCH_ASSOC);
+        $findOneStatement->setFetchMode(PDO::FETCH_CLASS|PDO::FETCH_PROPS_LATE, 'Phrases\Entity\Phrase', ['temp_title', 'temp_text']);
```
Comment (malukenho) [inline src/Phrases/Persistance/Relational.php]: Should be `Phrases\Entity\Phrase::class`

File: src/Phrases/Persistance/Relational.php [hunk 2]
```diff
@@ -42,7 +43,8 @@ protected function getPdo()
...
+        $onePhrase = $findOneStatement->fetch();
         if (empty($onePhrase)) {
             return [];
```
Comment (malukenho) [inline src/Phrases/Persistance/Relational.php]: Can be `return $findOneStatement->fetch()`

Comment (jackmakiyama) [inline src/Phrases/Persistance/Relational.php]: The application expected a array return.

File: src/Phrases/Persistance/Memory.php [hunk 3]
```diff
@@ -2,12 +2,22 @@
...
 
     public function __construct(array $list)
     {
+        foreach ($list as $phrase) {
+            if ($phrase instanceof Phrase) {
+                continue;
+            }
+
+            $message = 'Phrase list should contain only Phrase entities.';
+            throw new \InvalidArgumentException($message);
```
Comment (malukenho) [inline src/Phrases/Persistance/Memory.php]: We can avoid this one and craete a new strucure only accept `Phrase`, maybe a `PhraseCollection`.

File: tests/PhrasesTest/Persistance/AbstractRelationalTestCase.php [hunk 4]
```diff
@@ -106,14 +111,28 @@ public function testFindOneRandomWithPhrasesInTheTableReturnsFromTheList()
...
-            $firstPhrase = $phrases->findOneRandom(),
-            $list
+            $firstPhrase->getTitle(),
+            $this->existingTitles,
+            'First phrase should be on the list of phrases in the persistance storage.'
         );
 
+        $secondPhrase = $phrases->findOneRandom();
+        $this->assertInstanceOf(
+            'Phrases\\Entity\\Phrase',
```
Comment (malukenho) [inline tests/PhrasesTest/Persistance/AbstractRelationalTestCase.php]: `::class`

File: src/Phrases/Entity/Phrase.php [hunk 5]
```diff
@@ -9,31 +9,23 @@ class Phrase
     protected $title;
     protected $text;
 
-    public function setTitle($title)
+    public function __construct($title, $text)
     {
         if (empty($title)) {
```
Comment (malukenho) [inline src/Phrases/Entity/Phrase.php]: I think we need test for this.

Comment (malukenho) [inline src/Phrases/Entity/Phrase.php]: Also, use `trim` too.

File: src/Phrases/Persistance/Memory.php [hunk 6]
```diff
@@ -2,12 +2,22 @@
 
 namespace Phrases\Persistance;
 
+use Phrases\Entity\Phrase;
+
 class Memory implements RepositoryInterface
 {
     protected $list = array();
```
Comment (malukenho) [inline src/Phrases/Persistance/Memory.php]: short array syntax here

File: src/Phrases/Persistance/Memory.php [hunk 7]
```diff
@@ -25,8 +35,9 @@ public function findOneRandom()
         return $this->list[$random];
     }
 
-    public function save(array $phrase)
+    public function save(Phrase $phrase)
     {
-        $this->list[] = $phrase;
+        return $this->list[] = $phrase;
```
Comment (malukenho) [inline src/Phrases/Persistance/Memory.php]: Why we have a return on save method? that's correct?

File: tests/PhrasesTest/Controller/GetPhraseTest.php [hunk 8]
```diff
@@ -23,22 +24,27 @@ private function createMemoryPersistanceForPhraseArray(array $list)
...
     }
 
     public function testGetSinglePhraseWhenMoreThanOnePhraseExists()
     {
-        $expectedPhrase = 'Eu sou lindo';
-        $phraseList = [$expectedPhrase, 'Eu sou feio'];
+        $expectedPhrase = new Phrase('Eu sou lindo', 'Só que não!');
+        $phrase = new Phrase('Eu sou feio', 'Só que é!');
+        $phraseList = [$expectedPhrase, $phrase];
+        $existingTexts = array();
```
Comment (malukenho) [inline tests/PhrasesTest/Controller/GetPhraseTest.php]: short array syntax

File: tests/PhrasesTest/Controller/GetPhraseTest.php [hunk 9]
```diff
@@ -23,22 +24,27 @@ private function createMemoryPersistanceForPhraseArray(array $list)
 
     public function testGetSinglePhraseWhenOnlyOnePhraseExists()
     {
-        $expectedPhrase = 'Eu sou lindo';
+        $expectedPhrase = new Phrase('Eu sou lindo', 'Só que não!');
         $phrases = $this->createMemoryPersistanceForPhraseArray([$expectedPhrase]);
         $controller = new GetPhrase($phrases);
         $request = $this->createStubRequestObject();
         $response = $controller->execute($request);
         $this->assertInstanceOf('Zend\Http\Response', $response);
```
Comment (malukenho) [inline tests/PhrasesTest/Controller/GetPhraseTest.php]: `::class`

File: tests/PhrasesTest/Persistance/AbstractRelationalTestCase.php [hunk 10]
```diff
@@ -15,6 +16,7 @@
      * @var \Phrases\Persistance\Relational
      */
     protected $relational;
+    protected $existingTitles = array();
```
Comment (malukenho) [inline tests/PhrasesTest/Persistance/AbstractRelationalTestCase.php]: short array syntax

File: tests/PhrasesTest/Persistance/AbstractRelationalTestCase.php [hunk 11]
```diff
@@ -23,6 +25,12 @@ protected function setUp()
         $this->assertNotEmpty($this->relational, 'Please, set up a relational adapter on the test case.');
         $this->pdo->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
         $this->relationalProvider = ConsumedData::asRelationalArray();
+        $this->relationalEntityPhraseProvider = array();
```
Comment (malukenho) [inline tests/PhrasesTest/Persistance/AbstractRelationalTestCase.php]: short array syntax

File: tests/PhrasesTest/Persistance/AbstractRelationalTestCase.php [hunk 12]
```diff
@@ -106,14 +111,28 @@ public function testFindOneRandomWithPhrasesInTheTableReturnsFromTheList()
 
         $phrases = $this->relational;
 
+        $firstPhrase = $phrases->findOneRandom();
+        $this->assertInstanceOf(
+            'Phrases\\Entity\\Phrase',
```
Comment (malukenho) [inline tests/PhrasesTest/Persistance/AbstractRelationalTestCase.php]: `::class`

File: tests/PhrasesTest/Persistance/ArrayTest.php [hunk 13]
```diff
@@ -72,14 +79,26 @@ public function testFindOneRandomWithPhrasesInTheListReturnsFromTheList()
...
         );
     }
+
+    /**
+     * @expectedException InvalidArgumentException
+     * @expectedExceptionMessage Phrase list should contain only Phrase entities.
+     */
+    public function testMemoryInstantiatedWithArrayOfArraysThrowsAnException()
+    {
+        $phraseList = array(
```
Comment (malukenho) [inline tests/PhrasesTest/Persistance/ArrayTest.php]: array short syntax

**Comment by malukenho** [comment]
@jackmakiyama also, I don't see a test passing weird values for `Entity\Phrase`, as a `stdClass`, resource, or anything else.

File: src/Phrases/Persistance/Relational.php [hunk 14]
```diff
@@ -50,14 +52,15 @@ protected function getPdo()
         return $onePhrase;
```
Comment (malukenho) [inline src/Phrases/Persistance/Relational.php]: Why modify this behavior? It's commom to return anything or NIL

File: tests/PhrasesTest/ApplicationTest.php [hunk 15]
```diff
@@ -22,8 +23,8 @@ class ApplicationTest extends PHPUnit_Framework_TestCase
 
     public function setUp()
     {
-        $phrases = ConsumedData::asArray();
-        $onePhrase = array_shift($phrases);
+        $phrases = ConsumedData::asRelationalArray();
```
Comment (malukenho) [inline tests/PhrasesTest/ApplicationTest.php]: `ConsumedData::asRelationalArray()[0];`

File: tests/PhrasesTest/Controller/GetPhraseTest.php [hunk 16]
```diff
@@ -23,22 +24,27 @@ private function createMemoryPersistanceForPhraseArray(array $list)
...
 
     public function testGetSinglePhraseWhenMoreThanOnePhraseExists()
     {
-        $expectedPhrase = 'Eu sou lindo';
-        $phraseList = [$expectedPhrase, 'Eu sou feio'];
+        $expectedPhrase = new Phrase('Eu sou lindo', 'Só que não!');
+        $phrase = new Phrase('Eu sou feio', 'Só que é!');
+        $phraseList = [$expectedPhrase, $phrase];
+        $existingTexts = array();
+        foreach ($phraseList as $phrase) {
```
Comment (malukenho) [inline tests/PhrasesTest/Controller/GetPhraseTest.php]: `array_map`

File: tests/PhrasesTest/Controller/GetPhraseTest.php [hunk 17]
```diff
@@ -46,7 +52,7 @@ public function testGetSinglePhraseWhenMoreThanOnePhraseExists()
         $this->assertInstanceOf('Zend\Http\Response', $response);
```
Comment (malukenho) [inline tests/PhrasesTest/Controller/GetPhraseTest.php]: `::class`

File: tests/PhrasesTest/Persistance/ArrayTest.php [hunk 18]
```diff
@@ -2,11 +2,24 @@
...
 
+use Phrases\Entity\Phrase;
+use PhrasesTestAsset\ConsumedData;
+
 /**
  * @small
  */
 class MemoryTest extends \PHPUnit_Framework_TestCase
 {
+    protected $phraseList = array();
```
Comment (malukenho) [inline tests/PhrasesTest/Persistance/ArrayTest.php]: short syntax

**Comment by malukenho** [comment]
@jackmakiyama It's should be rebased.

File: src/Phrases/Entity/Phrase.php (hunk 5 above)
Comment (jackmakiyama) [inline src/Phrases/Entity/Phrase.php]: The tests are in this file: `tests/PhrasesTest/Entity/PhraseTest.php`

File: src/Phrases/Persistance/Memory.php (hunk 3 above)
Comment (jackmakiyama) [inline src/Phrases/Persistance/Memory.php]: I thought of that too.

File: src/Phrases/Persistance/Relational.php (hunk 14 above)
Comment (jackmakiyama) [inline src/Phrases/Persistance/Relational.php]: @malukenho  I did not understand exactly what you said.

File: tests/PhrasesTest/ApplicationTest.php (hunk 15 above)
Comment (jackmakiyama) [inline tests/PhrasesTest/ApplicationTest.php]: Oh, Thanks!

File: tests/PhrasesTest/Controller/GetPhraseTest.php (hunk 16 above)
Comment (jackmakiyama) [inline tests/PhrasesTest/Controller/GetPhraseTest.php]: @malukenho I do not know the right way to use it in this case =[

**Comment by malukenho** [comment]
@jackmakiyama why you close that?

**Comment by jackmakiyama** [comment]
See Skype please... =S

**Comment by jackmakiyama** [comment]
I pushed to your repository... :'(... I need you to push again yours commits. =S

**Pull request closed without merging**